### PR TITLE
Add BASNet to keras hub 

### PR DESCRIPTION
**Pull request by laxmareddyp**
Refer this [issue](https://github.com/keras-team/keras-hub/issues/1761) for more details

### CODE REVIEW

File: keras_hub/src/models/basnet/basnet_test.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    @pytest.mark.large
+    def test_saved_model(self):
+        self.run_model_saving_test(
+            cls=BASNet,
+            init_kwargs=self.init_kwargs,
+            input_data=self.images,
+        )
+
+    def test_end_to_end_model_predict(self):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/basnet/basnet_test.py:36]: add presets test and also self.run_task_test

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks for the PR Laxma! it looks great! can you add a demo colab to verify outputs? Thanks!

**Review by mattdangerw — COMMENTED** [review]
Started a review. But I think this might need some bigger considerations.

Some questions:
- What are the weight sources we are using here?
- What are the outputs? Looks like there is a list of outputs instead of a normal classification style output.
- How does training this model look?

In general, I think there there will be a lot of core infrastructure that will break if the arch does not have a backbone class. [DeepLabV3](https://github.com/keras-team/keras-hub/tree/master/keras_hub/src/models/deeplab_v3)  is probably much closer to what we will need to go for.

But I think I need answers to the above questions to better suggest a design.

**Comment by laxmareddyp** [comment]
> Some questions:
> 
> * What are the weight sources we are using here?

The weights conversion script uses the weights from https://keras.io/examples/vision/basnet_segmentation/ . There's  also a Keras BASNet model available on Kaggle, but it failed to load for me so far, and I don't know if where it comes from, if it's better or if it's the same model anyway.

> * What are the outputs? Looks like there is a list of outputs instead of a normal classification style output?

Yes, the direct model output is a list  for refinement and prediction  modules. The different stages of the model are trained with the same segmentation mask data. For inference, access to different stages of prediction modules would not be required for the user, but it might be interesting and doesn't hurt.


> * How does training this model look?

Training is simply performed on segmentation map data. For reference check this keras-cv [notebook](https://colab.sandbox.google.com/gist/chunduriv/27cd0f239d5184f88f1ca07a225059c0/basnet_segmentation.ipynb) published in keras-io and [issue](https://github.com/keras-team/keras-io/pull/1942)

File: keras_hub/src/models/basnet/basnet_test.py (hunk 1 above)
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet_test.py:36]: Done.

File: keras_hub/src/models/basnet/basnet.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+from keras_hub.src.api_export import keras_hub_export
+from keras_hub.src.models.basnet.basnet_preprocessor import BASNetPreprocessor
+from keras_hub.src.models.image_segmenter import ImageSegmenter
+from keras_hub.src.models.resnet.resnet_backbone import ResNetBackbone
+from keras_hub.src.models.resnet.resnet_backbone import (
+    apply_basic_block as resnet_basic_block,
+)
+
+
+@keras_hub_export(
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet.py:12]: Modified and will push the changes.

File: keras_hub/src/models/basnet/basnet.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    [
+        "keras_hub.models.BASNet",
+    ]
+)
+class BASNet(ImageSegmenter):
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet.py:17]: Done.

File: keras_hub/src/models/basnet/basnet.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+    # Train model
+    model.compile(
+        optimizer="adam",
+        loss=keras.losses.BinaryCrossentropy(from_logits=False),
+        metrics=["accuracy"],
+    )
+    model.fit(images, labels, epochs=3)
+        ```
+    """  # noqa: E501
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet.py:74]: Done

File: keras_hub/src/models/basnet/basnet.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+        # Refinement model.
+        refine_model = basnet_rrm(
+            predict_model, projection_filters, refinement_head
+        )
+
+        outputs = refine_model.outputs  # Combine outputs.
+        outputs.extend(predict_model.outputs)
+
+        outputs = [
+            keras.layers.Activation("sigmoid", dtype="float32")(_)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet.py:149]: Done( and added names for outputs)

File: keras_hub/src/models/basnet/basnet.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+    output = model(images)
+    pred_labels = output[0]
+
+    # Train model
+    model.compile(
+        optimizer="adam",
+        loss=keras.losses.BinaryCrossentropy(from_logits=False),
+        metrics=["accuracy"],
+    )
+    model.fit(images, labels, epochs=3)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet.py:72]: This 'fit()' does actually work (on my system), but I'll check again if it has the desired effect. We can alternatively implement a custom compute_loss() and avoid the multiple output model, if this is preferred. Shouldn't be hard imo.

File: keras_hub/src/models/basnet/basnet.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    segmentation.
+
+    References:
+        - [BASNet: Boundary-Aware Segmentation Network for Mobile and Web Applications](https://arxiv.org/abs/2101.04704)
+
+    Args:
+        backbone: `keras.Model`. The backbone network for the model that is
+            used as a feature extractor for BASNet prediction encoder. Currently
+            supported backbones are ResNet18 and ResNet34. Default backbone is
+            `keras_cv.models.ResNet34Backbone()`.
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet.py:29]: Done

**Comment by mattdangerw** [comment]
Maybe for now just directly subclass resnet without changes as `BasnetBackbone`? That way we can satisfy all the infra that will expect every new arch to have a backbone class.

The broader question I have is how to make our `ImageSegmenter` tasks interchangeable in terms of data in and prediction out. The goal of the task API is to allow users to swap different arches while keep the overall training/inference code the same. But it seems like there's a number of different ways to pose the segmentation problem. Not sure what to do here, but we can at least add a backbone class for now.

**Comment by mattdangerw** [comment]
Nice! Will take a look. Test failures look to be legit and basnet related, please take a look!

File: keras_hub/src/models/basnet/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from keras_hub.src.models.basnet.basnet_backbone import BASNetBackbone
+from keras_hub.src.models.basnet.basnet_presets import basnet_presets
+from keras_hub.src.utils.preset_utils import register_presets
+
+register_presets(basnet_presets, BASNetBackbone)
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/__init__.py:5]: newline at end of file

File: keras_hub/src/models/basnet/basnet.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+from keras_hub.src.api_export import keras_hub_export
+from keras_hub.src.models.basnet.basnet_backbone import BASNetBackbone
+from keras_hub.src.models.basnet.basnet_preprocessor import BASNetPreprocessor
+from keras_hub.src.models.image_segmenter import ImageSegmenter
+
+
+@keras_hub_export("keras_hub.models.BASNetImageSegmenter")
+class BASNetImageSegmenter(ImageSegmenter):
+    """
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet.py:11]: always have a short description for docstring that fits on a single line, immediately following the triple quote.

File: keras_hub/src/models/basnet/basnet.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        # only return the refinement module's output as final prediction
+        outputs = outputs["refine_out"]
+        super().__init__(inputs=x, outputs=outputs, **kwargs)
+
+        # === Config ===
+        self.backbone = backbone
+        self.preprocessor = preprocessor
+
+    def compute_loss(self, x, y, y_pred, *args, **kwargs):
+        # train BASNet's prediction and refinement module outputs against the same ground truth data
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet.py:77]: wrap at 80 characters. recent formatting changes will catch this if you sync with master

File: keras_hub/src/models/basnet/basnet.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        outputs = self.backbone(x)
+        losses = []
+        for output in outputs.values():
+            losses.append(super().compute_loss(x, y, output, *args, **kwargs))
+        loss = keras.ops.sum(losses, axis=0)
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet.py:82]: return immediately

File: keras_hub/src/models/basnet/basnet.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        if loss == "auto":
+            loss = keras.losses.BinaryCrossentropy()
+        if metrics == "auto":
+            metrics = [keras.metrics.Accuracy()]
+        super().compile(
+            optimizer=optimizer,
+            loss=loss,
+            metrics=metrics,
+            **kwargs,
+        )
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet.py:128]: newline at end of file

File: keras_hub/src/models/basnet/basnet_backbone.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+            projecting low-level features from the `backbone`.
+        prediction_heads: (Optional) List of `keras.layers.Layer` defining
+            the prediction module head for the model. If not provided, a
+            default head is created with a Conv2D layer followed by resizing.
+        refinement_head: (Optional) a `keras.layers.Layer` defining the
+            refinement module head for the model. If not provided, a default
+            head is created with a Conv2D layer.
+        dtype: `None` or str or `keras.mixed_precision.DTypePolicy`. The dtype
+            to use for the model's computations and weights.
+
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet_backbone.py:38]: remove empty line

File: keras_hub/src/models/basnet/basnet_backbone.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+        x = convolution_block(x, filters=filters)
+
+    x = segmentation_head(x)  # Refinement segmentation head.
+
+    # ------------- refined = coarse + residual
+    x = keras.layers.Add(dtype=dtype)(
+        [x_input, x]
+    )  # Add prediction + refinement output
+
+    return keras.models.Model(inputs=base_model.input, outputs=[x])
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet_backbone.py:370]: newline at end of file (will stop commenting this)

File: keras_hub/src/models/basnet/basnet_presets.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+basnet_presets = {
+    "basnet_resnet18": {
+        "metadata": {
+            "description": "BASNet with a ResNet18 v1 backbone.",
+            "params": 98780872,
+            "official_name": "BASNet",
+            "path": "basnet",
+        },
+        "kaggle_handle": "",  # TODO
```
Comment (mattdangerw) [inline keras_hub/src/models/basnet/basnet_presets.py:11]: either let's add these before we merge, or remove this entirely before our presets our ready

**Review by mattdangerw — COMMENTED** [review]
Let's try this out! I wonder if we will need to make more tweaks, having the num classes on the backbone itself is kinda funky, but do get why we are doing it for this architecture.

Left a number of style comments.

File: keras_hub/src/models/basnet/basnet_presets.py (hunk 15 above)
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet_presets.py:11]: Ok, i will remove this , later will add once presets are ready.

File: keras_hub/src/models/basnet/basnet_backbone.py (hunk 14 above)
Comment (laxmareddyp) [reply keras_hub/src/models/basnet/basnet_backbone.py:370]: I tried to run the shell commands to fix the format issues but i could see `ruff ` command not found as error. I did sync with master before pushing the code to PR but didn't work out the commands. I will figure it out. Thanks

**Comment by mattdangerw** [comment]
@laxmareddyp check the test failures! ~6 lines are too long and need fixing

**Comment by laxmareddyp** [comment]
> @laxmareddyp check the test failures! ~6 lines are too long and need fixing

@mattdangerw I am modifying the code related to doc strings, other minor suggestions you mentioned and will push the changes and hope it will not show up this time. Thanks

File: tools/checkpoint_conversion/convert_basnet_checkpoints.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python3
```
Comment (mattdangerw) [inline tools/checkpoint_conversion/convert_basnet_checkpoints.py:1]: I'll leave this off the PR for now.

File: tools/checkpoint_conversion/convert_basnet_checkpoints.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+Downloads BASNet modelweights with ResNet34 backbone and converts them to a
+KerasHub model. Credits for model training go to Hamid Ali
+(https://github.com/hamidriasat/BASNet).
+
+Requirements:
+pip3 install -q git+https://github.com/keras-team/keras-hub.git
+pip3 install -q gdown
+"""
+
+import gdown
```
Comment (mattdangerw) [inline tools/checkpoint_conversion/convert_basnet_checkpoints.py:16]: let's avoid a weird deps like this. We don't use it elsewhere. Should be able to make a bare link to a google drive for now, but more importantly this isn't converting anything, it's just wrapping some weights into a preset.

File: tools/checkpoint_conversion/convert_basnet_checkpoints.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    input_conv_kernel_sizes=[7],
+    stackwise_num_filters=[64, 128, 256, 512],
+    stackwise_num_blocks=[3, 4, 6, 3],
+    stackwise_num_strides=[1, 2, 2, 2],
+    block_type="basic_block",
+)
+
+# instantiate BASNet and load pretrained weights
+preprocessor = BASNetPreprocessor()
+backbone = BASNetBackbone(image_encoder=image_encoder, num_classes=1)
```
Comment (mattdangerw) [inline tools/checkpoint_conversion/convert_basnet_checkpoints.py:41]: Why only one class? What is this predicting.

File: tools/checkpoint_conversion/convert_basnet_checkpoints.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+import gdown
+
+from keras_hub.src.models.basnet.basnet import BASNetImageSegmenter
+from keras_hub.src.models.basnet.basnet_backbone import BASNetBackbone
+from keras_hub.src.models.basnet.basnet_preprocessor import BASNetPreprocessor
+from keras_hub.src.models.resnet.resnet_backbone import ResNetBackbone
+
+# download weights
+gdown.download(
+    "https://drive.google.com/uc?id=1OWKouuAQ7XpXZbWA3mmxDPrFGW71Axrg",
```
Comment (mattdangerw) [inline tools/checkpoint_conversion/convert_basnet_checkpoints.py:25]: What is this weight? Where's it from? How did we manipulate it?

**Review by mattdangerw — COMMENTED** [review]
Looks good overall. I'll drop the conversion script as I merge. Let's do that separately, need some work
